### ISSUE TITLE
Color rendering on PB741

### ISSUE THREAD

**Issue by GuillaumeSimonFr** [koreader/koreader#8872]
KOReader version: v2022.02
Device: PocketBook InkPad Color (PB741)

Since last version, color are working on PB741, but they remain very dull.
They are far more drab than with PBReader.

**Comment by Frenzie** [comment]
Pinging @rjd22 

I think that kind of sounds like maybe they up the contrast in the rendering stage?

**Comment by rjd22** [comment]
Hmm, I'm not sure it depends on if KoReader contrast setting also influences color contrast. There are no obvious ways of controlling contrast or color temperature in inkview.

@GuillaumeSimonFr did you try playing with the contrast setting in KoReader?

**Comment by Frenzie** [comment]
@rjd22 In PDF it does.

**Comment by GuillaumeSimonFr** [comment]
Yes I tried to adjust the contrast but the colors remain grayish

At 1, the rendering is very clear but colors are hard to see
At 5 and beyond, it's too dark
The default 2 seems to be a good compromise but remains too grayish.

Forgot to mention I did my test with pdf files.
I'll make some tests with other formats

**Comment by GuillaumeSimonFr** [comment]
Tried with an epub (Olivert Twist, delivered with the PB741) but it's grayish too

**Comment by rjd22** [comment]
I checked it on my pocketbook in a dark room and I can indeed see some difference in saturation between Pocketbook reader and KoReader. My guess is that it depends on what KoReader feeds inkview and there is quite little control on the image.

Pocketbook reader:
![IMG_20220304_190526](https://user-images.githubusercontent.com/160743/156819287-fe851189-9112-498b-8969-3bc30ded836c.jpg)

KoReader (contrast gamma 0.8)
![IMG_20220304_190617](https://user-images.githubusercontent.com/160743/156819379-80621e63-6f6f-46c9-9834-96c24eada854.jpg)

KoReader (contrast gamma 1.0)
![IMG_20220304_190821](https://user-images.githubusercontent.com/160743/156819459-738baf97-560c-4cda-a700-ae7ca0191357.jpg)

IMHO this is not really an issue with the Pocketbook implementation but more with KoReader handing inkview a less saturated image and not having any controls to make it warmer.

This is most likely because kaleido screens started working like 2 weeks ago for KoReader and there was no need yet to make controls for color temperature.

I need a second opinion from @Frenzie and @NiLuJe on this tho.

**Comment by Frenzie** [comment]
> IMHO this is not really an issue with the Pocketbook implementation but more with KoReader handing inkview a less saturated image and not having any controls to make it warmer.

We're not providing a less saturated image; we're providing the image as-is. If that doesn't look great we'd have to work around it by saturating it but imho that is 100% an inkview and/or kernel driver issue.

**Comment by NiLuJe** [comment]
I'd need an fbgrab dumb to (possibly?) confirm, but, IIRC, there's also a bit of dithering applied by InkView. What the reader app does before that is another kettle of fish: we tend to do nothing, as @Frenzie said.

**Comment by rjd22** [comment]
@NiLuJe @Frenzie Thank you for your input. To increase my understanding. At wat level are the contrast settings handled? If I understand correctly this is not done on the image but in a later stage?

Inkview most likely applies dithering. I'm not sure what version of it is used.

**Comment by Frenzie** [comment]
> At wat level are the contrast settings handled? If I understand correctly this is not done on the image but in a later stage?

Otoh the optional contrast adjustment in paged media is a part of k2pdfopt, but I'm not really sure which distinction you're trying to make by speaking of a later stage.

**Comment by rjd22** [comment]
> Otoh the optional contrast adjustment in paged media is a part of k2pdfopt, but I'm not really sure which distinction you're trying to make by speaking of a later stage.

It's kind of hard to convey what I mean because I'm still learning and don't have a complete understanding of every thing that happens between loading the image from the file system and rendering the on the screen.

What I'm wondering is if contrast is applied to the image before it's send to the inkview/framebuffer(am I using this term correctly?) or is it applied after it's send to inkview trough some setting?

Again sorry if I'm misunderstanding something.

**Comment by Frenzie** [comment]
I'm not quite sure I follow. The only way to do it "after" would be by passing some kind of `oversaturate = true` flag/bitmask along to inkview, if that's possible. (I assume you know more about inkview than me. ;-) ) In that case it would theoretically perform some transformations to the image itself prior to writing it to the framebuffer, or in a best case scenario optimized by hardware, all of which would be the final step from KOReader's perspective. Anything else can only be prior to that.

**Comment by Frenzie** [comment]
Basically what happens is:

* we ask crengine/MuPDF/djvulibre/whatever to draw a page with certain properties
* some transformations might optionally be applied, either directly in the engine as an  or on the result
* that's cached to a blitbuffer
* that's (generally) used as the background widget
* other widgets are drawn on top of that
* the final result is sent to the framebuffer (or SDL/Native Window/Inkview which function in a similar manner)
  * this might include some final pre-processing like SW/HW dithering as a kind of pre-final step, it could potentially make sense for saturation adjustment to live there too I suppose

**Comment by rjd22** [comment]
I've send PocketBook an email about this and got the following answer:

Pocketbook reader uses several algorithms to improve colors. 
​
There are the following functions in the interview library (declared in inkview.h): 
 
`void adjustBitmapDefault(ibitmap* bmp);
void adjustAreaDefault(unsigned char* data, int scanline, int width, int height);`

These functions adjust colors according to predefined parameters on Pocketbook device. These parameters could be modified using configuration file – /mnt/ext1/system/config/images_parameters.cfg. 
Default parameters are stored in /ebrmain/config/images_parameters.cfg

If more customized adjustment is needed, you could use functions with specific adjustment algorithm directly. 

For color device:
```
/**
 * @brief adjustBitmapWithSaturationBrightnessAndDithering - change saturation and value of bitmap pixels (as in HVS color representation)
 *        and optionally apply ordered dithering
 * @param bmp - bitmap to adjust
 * @param saturation_table - table with 256 elements each element - what substitute corresponding saturation with.
 * @param value_table - same as saturation table but fur value
 * @param dithering - apply dithering or not.
 */
void adjustBitmapWithSaturationBrightnessAndDithering(ibitmap* bmp, int * saturation_table, int * value_table, bool dithering);
void adjustAreaWithSaturationBrightnessAndDithering(unsigned char* data, int scanline, int width, int height, int * saturation_table,
                                                    int * value_table, bool dithering);
```

This function changes saturation using LAB color model:

`void adjustAreaWithLabColorsSaturation(unsigned char* data, int scanline, int width, int height, double koef);`

This function changes image vibrance:

`void adjustAreaWithVibrance(unsigned char* data, int scanline, int width, int height, float vibrance);`

This function performs quick but low quality adjustment (we are using it for A2 mode):

`void ditherAndAdjustArea(unsigned char* data, int scanline, int width, int height, AdjustmentParams* params);`


For grayscale device the next function changes brightness, contrast and gamma:

`void adjustAreaWithBrightnessContrastGamma(unsigned char * data, int scanline, int width, int height, double brightness, double contrast, double gamma);`

**Comment by rjd22** [comment]
@NiLuJe @Frenzie I've been playing with the `adjustAreaWithVibrance` with really great results. Being able to adjust this in the settings would really improve the reading experience for users with kaleido screens.

What I'm wondering is what would be the correct way to implement this to allow users to adjust this so they can decide what looks best for them. Is there a way to implement device specific sliders?

**Comment by Frenzie** [comment]
> Is there a way to implement device specific sliders?

The same way as any other, just with a feature check. (A *device* check is rarely if ever desirable.)

**Comment by grabi678** [comment]
@rjd22  
> I've been playing with the adjustAreaWithVibrance with really great results. 

As long as there is no KOReader solution available for color improvement, could you please describe the steps you have been done while testing the color parameters? I tried to but could not even find the two config files on my Inkpad Color mentioned by Pocketbook:

> These parameters could be modified using configuration file – /mnt/ext1/system/config/images_parameters.cfg.
> Default parameters are stored in /ebrmain/config/images_parameters.cfg

This would be great. Many thanks.

**Comment by rjd22** [comment]
@grabi678 i didn't forget you but I'm really busy with other projects. When I have some time I will make a sample PR that shows how it works.
